packages/vouching/test/contracts/ZEPToken.test.js

```diff
@@ -37,10 +46,122 @@ contract('ZEPToken', ([ _, owner, another, jurisdiction]) => {
   })
 
   it('can be paused by creator', async function () {
-    await this.zepToken.pause({ from: owner });
+    await this.zepToken.pause({ from: tokenOwner });
   })
 
   it('cannot be paused by anybody', async function () {
     await assertRevert(this.zepToken.pause({ from: another }));
   })
+
+  describe('TPL', function () {
+    const amount = '5e18'
+    const sender = tokenOwner
+    const recipient = another
+
+    beforeEach('initialize and approve validator', async function () {
+      this.validator = await ZEPValidator.new()
+      const initializeValidatorData = encodeCall('initialize', ['address', 'uint256'], [this.jurisdiction.address, receiveTokensAttributeID])
+      await this.validator.sendTransaction({ data: initializeValidatorData, from: validatorOwner })
+
+      await this.jurisdiction.addValidator(this.validator.address, "ZEP Validator", { from: jurisdictionOwner })
+      await this.jurisdiction.addAttributeType(receiveTokensAttributeID, false, false, ZERO_ADDRESS, 0, 0, 0, "can transfer", { from: jurisdictionOwner })
+      await this.jurisdiction.addValidatorApproval(this.validator.address, receiveTokensAttributeID, { from: jurisdictionOwner })
+      await this.validator.addOrganization(zeppelin, 100, "ZEP Org", { from: validatorOwner })
+    })
+
+    function assertTokensCannotBeTransferred() {
+      it('cannot transfer', async function () {
+        assert.equal(await this.zepToken.canTransfer(recipient, amount, { from: sender }), false)
+        await assertRevert(this.zepToken.transfer(recipient, amount, { from: sender }))
+      })
+
+      it('cannot transfer from', async function () {
+        await this.zepToken.approve(recipient, amount, { from: sender })
+
+        assert.equal(await this.zepToken.canTransferFrom(sender, recipient, amount, { from: recipient }), false)
+        await assertRevert(this.zepToken.transferFrom(sender, recipient, amount, { from: recipient }))
+      })
+    }
+
+    function assertTokensCanBeTransferred() {
+      it('can transfer', async function () {
+        assert(await this.zepToken.canTransfer(recipient, amount, { from: sender }))
+        await this.zepToken.transfer(recipient, amount, { from: sender })
+
+        assert((await this.zepToken.balanceOf(recipient)).eq(amount))
+      })
+
+      it('can transfer from', async function () {
+        await this.zepToken.approve(recipient, amount, { from: sender })
+
+        assert(await this.zepToken.canTransferFrom(sender, recipient, amount, { from: recipient }))
+        await this.zepToken.transferFrom(sender, recipient, amount, { from: recipient })
+
+        assert((await this.zepToken.balanceOf(recipient)).eq(amount))
+      })
+    }
+
+    describe('when the no one was not allowed to receive tokens', function () {
```

Comment:
"the no one" sounds like a creature out of H P Lovecraft :-P

Reply:
bad copy paste 😂